v1alpha2 API Implementation

QuentinBisson opened the pull request:
As requested here https://github.com/giantswarm/silence-operator/pull/226#issuecomment-1259645330, this makes the silence CRD namespaced scope.

## 🎯 Overview

This PR introduces a new namespace-scoped `observability.giantswarm.io/v1alpha2` Silence API alongside the existing cluster-scoped `monitoring.giantswarm.io/v1alpha1`, adds a dedicated controller for v1alpha2, and provides migration tooling and documentation.

## 📋 Summary of Changes

### 🚀 New v1alpha2 API

- **Namespace-scoped**: Better multi-tenancy and RBAC isolation compared to cluster-scoped v1alpha1
- **Enhanced Validation**: Field validation with size limits (names 1-256 chars, values max 1024 chars)
- **MatchType Enum**: Replaces boolean fields with intuitive Alertmanager operator symbols (`=`, `!=`, `=~`, `!~`)
- **Cleaner API**: Removed deprecated fields (`targetTags`, `owner`, `postmortem_url`, `issue_url`)

### 🔧 Architecture Improvements

- **Service Layer**: Extracted business logic into `pkg/service/SilenceService` for better testability
- **Dual Controllers**: `SilenceV2Reconciler` for v1alpha2, sharing service logic with v1alpha1 controller
- **Dependency Injection**: Constructor-based service injection pattern

### 📋 API Comparison

| Feature | v1alpha1 | v1alpha2 |
|---------|----------|----------|
| Scope | Cluster-scoped | Namespace-scoped |
| API Group | `monitoring.giantswarm.io` | `observability.giantswarm.io` |
| Matcher Fields | `isRegex: bool, isEqual: bool` | `matchType: enum` |
| Validation | Basic | Enhanced with size limits |
| Finalizer | `monitoring.giantswarm.io/silence-protection` | `observability.giantswarm.io/silence-protection` |

### 🛠️ Migration Support

**Migration Script** (`hack/migrate-silences.sh`):
- Automated conversion from v1alpha1 to v1alpha2 format
- Boolean-to-enum conversion (`isRegex`/`isEqual` → `matchType`)
- System metadata filtering (preserves user annotations, removes K8s/Flux metadata)
- Dry-run support for safe testing
- Built-in help with examples

```bash
# Test migration without applying changes
./hack/migrate-silences.sh --dry-run

# Migrate to specific namespace
./hack/migrate-silences.sh production
```

### 📝 Documentation

- **MIGRATION.md**: Comprehensive migration guide with examples
- **Enhanced README.md**: Updated with v1alpha2 API documentation
- **Improved CHANGELOG.md**: Detailed feature descriptions

## 🎯 Benefits

- **Better Multi-tenancy**: Namespace-scoped resources improve security isolation
- **Cleaner API**: Intuitive `matchType: "="` vs complex boolean combinations
- **Enhanced Validation**: Prevents invalid configurations at API level
- **Gradual Migration**: Both APIs work simultaneously during transition
- **Future-Ready**: Foundation for continued API evolution
---

**Migration Note**: v1alpha1 API remains fully functional. New deployments should use v1alpha2 for improved features and security.

QuantumEnigmaa reviewed (approved):
LGTM once we will have decided how to handle the move to the new release with this breaking change

QuentinBisson commented:
On golem: 
![image](https://github.com/user-attachments/assets/3e7ac645-3bd8-412b-ae7e-0a8816a3759b)

QuentinBisson commented on internal/controller/silence_controller.go:
```diff
@@ -148,7 +155,7 @@ func (r *SilenceReconciler) reconcileCreate(ctx context.Context, silence *v1alph
 }
 
-// reconcileDelete handles the deletion of the external Alertmanager silence.
+// reconcileDelete handles the deletion of the external Alertmanager silence using the business logic service
```
```suggestion
// reconcileDelete handles the deletion of the external Alertmanager silence.
```

QuentinBisson commented on internal/controller/silence_v2_controller.go:
```diff
@@ -0,0 +1,188 @@
+	client.Client
+	Scheme *runtime.Scheme
+
+	Alertmanager   *alertmanager.AlertManager
```
```suggestion
```

QuentinBisson commented on pkg/alertmanager/alertmanager.go:
```diff
@@ -28,6 +28,20 @@ var (
+// Ensure Alertmanager implements Client
+var _ Client = (*Alertmanager)(nil)
+
+
```
```suggestion
```

TheoBrigitte commented on README.md:
```diff
@@ -59,21 +69,56 @@ helm upgrade [RELEASE_NAME] giantswarm/silence-operator
+
+**Key differences in v1alpha2:**
+- **Namespace-scoped**: Silences are scoped to specific namespaces instead of cluster-wide
+- **Enhanced status**: Comprehensive status tracking with conditions and phase information
```
Is this true ?

QuentinBisson replied:
I think it's some wrong context here

TheoBrigitte reviewed (commented):
`v1alpha2` feels strange as we are introducing a new api group anyway.

QuentinBisson commented on README.md:
```diff
@@ -59,21 +69,56 @@ helm upgrade [RELEASE_NAME] giantswarm/silence-operator
+2. Create equivalent v1alpha2 Silence resources in appropriate namespaces
+3. Verify the new silences are working correctly
+4. Remove old v1alpha1 resources
+5. Eventually remove the v1alpha1 CRD when no longer needed
```
Eventually means when we removed it from the operator

QuentinBisson commented:
@TheoBrigitte do you think having another v1alpha1 would help? I'm not sure

QuentinBisson commented on MIGRATION.md:
```diff
@@ -0,0 +1,462 @@
+echo "Next steps:"
+echo "1. Verify the migrated silences: kubectl get silences.observability.giantswarm.io -n $TARGET_NAMESPACE"
+echo "2. Test that silences work as expected"
+echo "3. Remove old v1alpha1 silences when confident: kubectl delete silences.monitoring.giantswarm.io <name>"
```
```suggestion
echo "3. Remove all old v1alpha1 silences when confident: kubectl delete silences.monitoring.giantswarm.io --all"
```

QuentinBisson commented:
I'll address your questions tomorrow and we can find some time to pair on it thursday or next week to get this over the finish line :)

QuentinBisson commented:
@TheoBrigitte I moved some of the files to another PR https://github.com/giantswarm/silence-operator/pull/566

This should only contain the new API and I have addressed all your feedback :)

Regarding the new API, I think this might be easier for existing users to have a new api version (not just a group) because in a yaml file, v1alpha1/v1alpha2 would make a better difference that observability or monitoring in the api group. WDYT?

QuentinBisson commented on CHANGELOG.md:
```diff
@@ -9,15 +9,31 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+### Deprecated
+
+- The `monitoring.giantswarm.io/v1alpha1` API is now considered legacy. New deployments should use `observability.giantswarm.io/v1alpha2`.
+- Cluster-scoped silences are deprecated in favor of namespace-scoped resources for improved security and multi-tenancy.
```
```suggestion
```

QuentinBisson commented on hack/migrate-silences.sh:
```diff
@@ -0,0 +1,247 @@
+    local isRegex
+    local isEqual
+    
+    name=$(echo "$matcher_json" | jq -r '.name')
+    value=$(echo "$matcher_json" | jq -r '.value')
+    isRegex=$(echo "$matcher_json" | jq -r '.isRegex // false')
+    isEqual=$(echo "$matcher_json" | jq -r 'if has("isEqual") then .isEqual else true end')
```
We need the has function because isEqual not set defaults to true (remember the weird boolean pointer in the v1alpha1 cr)

QuentinBisson commented on MIGRATION.md:
````diff
@@ -0,0 +1,445 @@
+
+#### Step 3: Verify v1alpha2 Silences Work
+
+```bash
+kubectl get silences.observability.giantswarm.io -n production
+kubectl describe silence example-silence -n production
````
Well, considering the CRDs do not have a status yet, nothing apart from checking them on alertmanager. What do you want me to add here?

TheoBrigitte commented on hack/migrate-silences.sh:
```diff
@@ -0,0 +1,247 @@
+        
+        # Log conversion
+        matcher_name=$(echo "$original_matcher" | jq -r '.name')
+        old_isRegex=$(echo "$original_matcher" | jq -r '.isRegex // false')
+        old_isEqual=$(echo "$original_matcher" | jq -r 'if has("isEqual") then .isEqual else true end')
```
I think displaying the raw input just makes more sense than changing it

QuentinBisson commented on MIGRATION.md:
```diff
@@ -0,0 +1,442 @@
+
+#### Step 3: Verify v1alpha2 Silences Work
+
+The best way to make sure the new silence works is to verify it was created on Alertmanager.
```
New ones will bé named silence-operator-namespace-silence_name while v1 silences are named silence-operator_silence_name. V1alpha2 will have the namespace in the name

QuentinBisson replied:
Also, v1alpha2 is not meant to stay long :)

hervenicol replied:
Maybe say so then:
```suggestion
The best way to make sure the new silence works is to verify it was created on Alertmanager. The new one contains the namespace in its name.
```

QuentinBisson commented on MIGRATION.md:
```diff
@@ -0,0 +1,442 @@
+kubectl get silences.monitoring.giantswarm.io
+kubectl get silences.observability.giantswarm.io --all-namespaces
+
+# Verify finalizers are properly set
+kubectl get silence test-v1alpha1 -o jsonpath='{.metadata.finalizers}'
+kubectl get silence test-v1alpha2 -n default -o jsonpath='{.metadata.finalizers}'
```
Hmm no it does not, we need to specify the group 👍🏻

QuentinBisson replied:
```suggestion
# Verify finalizers are properly set
kubectl get silences.monitoring.giantswarm.io test-v1alpha1 -o jsonpath='{.metadata.finalizers}'
kubectl get silences.observability.giantswarm.io test-v1alpha2 -n default -o jsonpath='{.metadata.finalizers}'
```

hervenicol commented on MIGRATION.md:
```diff
@@ -0,0 +1,442 @@
+
+---
+
+**Note**: Both API versions can coexist safely. The migration can be performed gradually without service interruption.
```
Should we, at some point, once all silences are migrated to v2, remove the v1 CRDs?

QuentinBisson replied:
This is the goal yes

QuentinBisson replied:
All new features will only bé for v1alpha2 as well

QuentinBisson replied:
No because the operator would fail to work

hervenicol commented on hack/migrate-silences.sh:
```diff
@@ -0,0 +1,253 @@
+      # Enable dry run mode
+      DRY_RUN=true;;
+    -?*)
+      echo "❌ Unknown option $1" >&2
+      exit 1;;
```
If user provided wrong options, he may need to have a look at the help.
```suggestion
      echo "❌ Unknown option $1" >&2
      show_help
      exit 1;;
```

hervenicol commented on MIGRATION.md:
```diff
@@ -0,0 +1,442 @@
+| **Scope** | Cluster-scoped | Namespace-scoped |
+| **Matcher Fields** | `isRegex: bool`, `isEqual: bool` | `matchType: string` (enum) |
+| **Validation** | Basic validation | Enhanced validation with field size limits |
+| **Deprecated Fields** | Includes `targetTags`, `owner`, `issue_url`, `postmortem_url` | Cleaned up, only essential fields |
```
I didn't even know we had these :sweat_smile: 
Basically, we are supposed to have these in annotations or labels, right? I think that's what we are actually doing.

hervenicol commented on hack/migrate-silences.sh:
```diff
@@ -0,0 +1,253 @@
+
+Migrates v1alpha1 silences to v1alpha2 format with automatic matchers conversion.
+
+Arguments:
+  target-namespace    Target namespace for v1alpha2 silences (default: default)
+  --dry-run           Show what would be migrated without creating resources
+  --help              Show this help message
```
The first one (make target-namespace mandatory) looks easier and good enough.

hervenicol commented on hack/migrate-silences.sh:
```diff
@@ -0,0 +1,296 @@
+show_help() {
+    local bin="$(basename "$0")"
+    cat << EOF
+<<<<<<< Updated upstream
```
conflicts remainings?

QuentinBisson replied:
Nope :D

QuentinBisson commented on hack/migrate-silences.sh:
```diff
@@ -0,0 +1,296 @@
+show_help() {
+    local bin="$(basename "$0")"
+    cat << EOF
+<<<<<<< Updated upstream
+Usage: $bin [target-namespace] [--dry-run|--help]
+=======
+Usage: $0 <target-namespace> [--dry-run|--help]
+>>>>>>> Stashed changes
```
```suggestion
Usage: $0 <target-namespace> [--dry-run|--help]
```

QuentinBisson commented on hack/migrate-silences.sh:
```diff
@@ -0,0 +1,253 @@
+    if [[ "${1:-}" == "--dry-run" ]]; then
+        TARGET_NAMESPACE=""
+    fi
+fi
```
Man I trusted you :D

QuentinBisson replied:
It's fixed :)

hervenicol reviewed (approved):
LGTM
